keps/sig-scheduling/2891-simplified-config/README.md

```diff
@@ -0,0 +1,861 @@
+
+[**Option 1**](#Option-1)
+
+This option describes defining a new plugin interface, which plugin developers will need to 
+implement in order for their plugin to take advantage of automated multi-point registration. 
+The intent of this is to provide plugin developers with tunable control over their plugin's 
+implementation and maintain semantic consistency with existing extension points. The drawbacks 
+to this option are that it requires every plugin to opt-in to take advantage and a more complex 
+code structure.
+
+[**Option 2**](#Option-2)
+
+This option defines the `MultiPoint` extension as transparently applicable to all plugins, using 
+type casting to automatically initialize every plugin for every extension point they implement 
+by checking each one. The benefit to this approach is that all plugins are capable of being registered 
+this way without any additional plugin code. The drawbacks of this are less defined control for 
+developers and minor changes to config semantics for users.
+
+
+#### Option 1
+* Define a new `MultiPointPlugin` interface in [`pkg/scheduler/framework/interface.go`](https://github.com/kubernetes/kubernetes/blob/dc079ac/pkg/scheduler/framework/interface.go):
+```go
+type MultiPointPlugin interface {
+  Plugin
+  // Register allows a plugin to automatically register itself in 
+  // multiple extension points, as defined by the plugin
+  Register(fw *framework.Framework)
+}
+```
+
+* Modify [`frameworkImpl`](https://github.com/kubernetes/kubernetes/blob/dc079acc2be603fbc2065e2748c79bb5e2c3453d/pkg/scheduler/framework/runtime/framework.go#L78-L88) to have a `multiPointPlugins` field and store the `pluginsMap` generated in [`NewFramework()`](https://github.com/kubernetes/kubernetes/blob/dc079acc2be603fbc2065e2748c79bb5e2c3453d/pkg/scheduler/framework/runtime/framework.go#L325)
+```go
+// frameworkImpl is the component responsible for initializing and running scheduler
+// plugins.
+type frameworkImpl struct {
+...
+	pluginsMap           map[string]config.Plugin // <--new
+	multiPointPlugins    []framework.MultiPointPlugin // <--new
+	scorePluginWeight    map[string]int
+	queueSortPlugins     []framework.QueueSortPlugin
+	preFilterPlugins     []framework.PreFilterPlugin
+	filterPlugins        []framework.FilterPlugin
+	postFilterPlugins    []framework.PostFilterPlugin
+	preScorePlugins      []framework.PreScorePlugin
+	scorePlugins         []framework.ScorePlugin
+	reservePlugins       []framework.ReservePlugin
+	preBindPlugins       []framework.PreBindPlugin
+	bindPlugins          []framework.BindPlugin
+	postBindPlugins      []framework.PostBindPlugin
+	permitPlugins        []framework.PermitPlugin
+...
+}
+```
+
+* Modify [`getExtensionPoints()`](https://github.com/kubernetes/kubernetes/blob/dc079acc2be603fbc2065e2748c79bb5e2c3453d/pkg/scheduler/framework/runtime/framework.go#L119) to return a the `MultiPoint` extensions as well, so that the factory functions for these get registered
```

Comment:
This KEP is going into the weeds by including too much code. It should definitely go away once the option is moved to "Alternatives".

Reply:
I specifically included very detailed code to show the exact steps each implementation will require in order to function as designed. I added more in the last update to answer some questions about how aspects like API defaulting will work. But I agree that once we have answered all questions and move forward with an implementation that this can definitely be cleaned up. I just didn't want to leave any "surprises" 🙂